### PR TITLE
Track closure signatures & kinds in freshened types

### PR DESCRIPTION
**Pull request by arielb1**
r? @nikomatsakis

### CODE REVIEW

File: src/librustc/infer/freshen.rs [hunk 1]
```diff
@@ -150,6 +157,53 @@ impl<'a, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for TypeFreshener<'a, 'gcx, 'tcx> {
...
+                    Some(ty::ClosureKind::FnMut) => tcx.types.i32,
+                    Some(ty::ClosureKind::FnOnce) => tcx.types.i64,
+                };
+
+                let params: AccumulateVec<[_; 8]> =
+                    substs.substs.iter().map(|k| k.fold_with(self))
+                    .chain([closure_sig_marker, closure_kind_marker].iter().cloned()
+                           .map(From::from))
+                    .collect();
+                let params = tcx.intern_substs(&params);
```
Comment (Mark-Simulacrum) [inline src/librustc/infer/freshen.rs]: Should be possible to call `mk_substs` on the iterator here.

File: src/librustc/infer/freshen.rs [hunk 2]
```diff
@@ -150,6 +157,53 @@ impl<'a, 'gcx, 'tcx> TypeFolder<'gcx, 'tcx> for TypeFreshener<'a, 'gcx, 'tcx> {
                 t
             }
 
+            ty::TyClosure(def_id, substs) => {
```
Comment (nikomatsakis) [inline src/librustc/infer/freshen.rs]: feels like some comments laying out the strategy here would be very welcome

**Review by nikomatsakis — APPROVED** [review]
OK, I get the idea here -- this is clever. The other approach that I had been contemplating was slightly different. I had hoped to add facts to the environment indicating when we know the kind of a closure (rather than invoking `closure_kind`, basically). In this way, we would invalidate the cached entries because there would be a new environment. But I think this is a simpler thing to do for now for sure (and maybe forever -- since I was worried about screwing up the caching results of other things if we grow the param env).

**Comment by arielb1** [comment]
Now with cleaner code & comments & hopefully ready for landing!

**Comment by arielb1** [comment]
Rebased

File: src/librustc/infer/freshen.rs [hunk 3]
```diff
@@ -19,10 +19,21 @@
...
 //! Note that you should be careful not to allow the output of freshening to leak to the user in
 //! error messages or in any other form. Freshening is only really useful as an internal detail.
 //!
-//! __An important detail concerning regions.__ The freshener also replaces *all* regions with
+//! Because of the manipulation required to handle closures, doing arbitrary operations on
+//! freshened types is not recommended. However, in addition to doing equality/hash
+//! comparisons (for caching), it is possible to do a `ty::_match` operation between
+//! 2 freshened types - this works even with the closure encoding.
+//!
+//! __An important detail concerning regions.__ The freshener also replaces *all* free regions with
```
Comment (nikomatsakis) [inline src/librustc/infer/freshen.rs]: nice comments

**Review by nikomatsakis — APPROVED** [review]

**Comment by arielb1** [comment]
- #44159 (I think - this *looks* spurious)

@bors retry

**Comment by arielb1** [comment]
Please don't roll this up as this is a performance improvement PR.

**Pull request merged**